README.md

```diff
@@ -18,68 +18,98 @@ As a part of the bridge, a Kafka producer will be responsible for producing mess
 This Kafka topic is defined by a template, `Kafka Topic template`.
 However, if the incoming MQTT message's topic does not match any pattern in the ToMaR, the Bridge has a default Kafka topic where the incoming message will be mapped to, known as `messages_default`.
 
-A valid ToMaR is a JSON file that contains an array of mapping rules. 
+The optional `Kafka record key` is used to define the key of the Kafka record that will be produced to the Kafka topic.
+It is defined by a template as well, and its default value is `null`.
+A valid ToMaR is a JSON file that contains an array of mapping rules.
 Each mapping rule is a JSON object that contains two properties: `mqttTopic` and `kafkaTopic`.
-The following is an example of a valid TOMAR:
+The following is an example of a valid ToMaR:
 
 ```json
 [
   {
-    "mqttTopic": "building/{building}/room/{room}/#",
-    "kafkaTopic": "building_{building}_room_{room}"
+     "mqttTopic": "building/(\\w+)/room/(\\d{1,4})/.*",
+     "kafkaTopic": "building_$1",
+     "kafkaKey": "room_$2"
   },
   {
-    "mqttTopic": "sensors/+/data",
+    "mqttTopic": "sensors/([^/]+)/data",
     "kafkaTopic": "sensor_data"
   },
   {
-    "mqttTopic": "sensors/#",
-    "kafkaTopic": "sensor_others"
+     "mqttTopic": "sensors.*",
+     "kafkaTopic": "sensor_others"
+  },
+  {
+    "mqttTopic": "devices/([^/]+)/data/(\b(all|new|old)\b)",
+     "kafkaTopic": "device_$1_data",
+     "kafkaKey": "device_$2"
+  },
+  {
+      "mqttTopic": "locations/([^/]+)(?:\\/.*)?$",
+      "kafkaTopic": "locations",
+      "kafkaKey": "location_$1"
   }
 ]
 ```
 
-The wildcard `#` represents one or more levels in the MQTT topic hierarchy.
-The wildcard `+` represents a single level in the MQTT topic hierarchy.
+The expressions `.*` and `(?:\\/.*)?$` are used to represent the wildcard `#`, which in turn represents one or more levels in the MQTT topic hierarchy.
+However, there are some cases that can lead to unexpected behavior when using the these wildcards interchangeably . 
+Therefore, you should note the following:
+  - You cannot use the `.*` wildcard in capturing groups. 
+  For example, the pattern `building/(\\w+)/room/(\\d{1,4})/(.*)` is invalid because it will capture the whole subtopic of the pattern, and `$3` placeholder might include `/` characters, which will lead to an invalid Kafka topic name.
+  - You can use the `.*` wildcard without a preceding `/` character, but it is not equivalent to when it is preceded by `/`. 
+  For example, the pattern `building.*` will match `building`, `building/room`, `buildingroom`, and so on. 
+  On the other hand, the pattern `building/.*` will match `building/room`, `building/floor/room`, and so on, but not `buildingroom` and `building/`. 
+  - You can also use the `(?:\\/.*)?$` wildcard to match the whole subtopic level of the pattern. 
+  For example, the pattern `locations/([^/]+)(?:\\/.*)?$` will match `locations/city/luanda/angola`, `locations/city`, `locations/city/`, and so on. 
```

Comment:
If you want to make a good comparison between `.*` and `(?:\\/.*)?$` you should use the same example and show how each of them is able to match differently.